core/src/main/java/io/grpc/ManagedChannelBuilderProvider.java

```diff
@@ -0,0 +1,94 @@
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ServiceLoader;
+
+/**
+ * Provider of managed channels for transport agnostic consumption.
+ *
+ * <p>Implementations <em>should not</em> throw. If they do, it may interrupt class loading. If
+ * exceptions may reasonably occur for implementation-specific reasons, implementations should
+ * generally handle the exception gracefully and return {@code false} from {@link #isAvailable()}.
+ */
+@Internal
+public abstract class ManagedChannelBuilderProvider {
+  private static final ManagedChannelBuilderProvider provider;
+
+  static {
+    ServiceLoader<ManagedChannelBuilderProvider> providers
+        = ServiceLoader.load(ManagedChannelBuilderProvider.class);
+    Iterator<ManagedChannelBuilderProvider> iter = providers.iterator();
+    List<ManagedChannelBuilderProvider> list = new ArrayList<ManagedChannelBuilderProvider>();
+    for (ManagedChannelBuilderProvider current : providers) {
+      if (!current.isAvailable()) {
+        continue;
+      }
+      list.add(current);
+    }
+    if (list.isEmpty()) {
+      provider = null;
+    } else {
+      provider = Collections.max(list, new Comparator<ManagedChannelBuilderProvider>() {
+        @Override
+        public int compare(ManagedChannelBuilderProvider f1, ManagedChannelBuilderProvider f2) {
+          return f1.priority() - f2.priority();
+        }
+      });
+    }
+  }
+
+  public static ManagedChannelBuilderProvider provider() {
```

Comment:
The SPIs I looked at didn't include "default" in the name. I saw things like "availableProviders" (note plural). The naming style used here is similar to that of nio, like in java.nio.channels.spi.